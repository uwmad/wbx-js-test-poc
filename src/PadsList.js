import React from 'react';
import PropTypes from 'prop-types';
import WebixHelper from './webix/WebixHelper';
import Dropdown from "./Dropdown";
import uuidv4 from 'uuid/v4';

const{ webix } = window;

/*eslint-disable func-names*/  //Functions passed in to webix are not arrow functions so that 'this' refers to webix instance.
/*eslint-disable react/no-this-in-sfc*/  //'this' is used to refer to webix in the handlers
/*eslint-disable arrow-body-style */  //leaving this so that our code looks more like the webix docs
/*eslint-disable react/self-closing-comp */ //two components contents are populated with vanilla js

const PadsList = (props) => {

    const { padsData, pagination, changePagination, shouldPaginate, height} = props;

    /* Stuff Related to Pagination Behavior */
    const pagerId = `PadsList_Pager_${uuidv4()}`;
    const pagerValues = [10,20,30,40,50,100];

    const pagerHtmlTemplate = `{common.first()}{common.prev()}<span class="spacer"></span>{common.next()}{common.last()}`;

    const pagerConfig = {
            size: pagination,
            template : pagerHtmlTemplate,
            container : pagerId,
            on : { 
                onAfterRender(){
                    
                    const {count, size, page, limit } = this.data;
                    const pagerNode = this.$view.parentNode;
                    const pagerBarNode = pagerNode.parentNode;

                    if(pagerBarNode){

                        const firstButton = pagerBarNode.querySelector('[webix_p_id="first"]') || {};
                        const prevButton  = pagerBarNode.querySelector('[webix_p_id="prev"]') || {};
                        const nextButton  = pagerBarNode.querySelector('[webix_p_id="next"]') || {};
                        const lastButton  = pagerBarNode.querySelector('[webix_p_id="last"]') || {};

                        const positionInfoNode = pagerBarNode.querySelector('.pagesListInfo');

                        const start = Math.min(  ( page * size ) + 1 , count );
                        const end = Math.min( ( page + 1 ) * size , count ); 

                        if(count <= size){

                            pagerBarNode.classList.add('disabled-pager');
                            firstButton.disabled = true;
                            prevButton.disabled = true;
                            nextButton.disabled = true;
                            lastButton.disabled = true;

                        }else{

                            pagerBarNode.classList.remove('disabled-pager');

                            if ( page === 0){

                                firstButton.disabled = true;
                                prevButton.disabled = true;
                                nextButton.disabled = false;
                                lastButton.disabled = false;

                            } else if (page === limit - 1){

                                firstButton.disabled = false;
                                prevButton.disabled = false;
                                nextButton.disabled = true;
                                lastButton.disabled = true;

                            } else {

                                firstButton.disabled = false;
                                prevButton.disabled = false;
                                nextButton.disabled = false;
                                lastButton.disabled = false;

                            }

                        }//check limits

                        if(positionInfoNode){
                            positionInfoNode.innerHTML = `${start} to ${end} of ${count} `; //Info Text
                        }

                    }//pagerBarNode exists

                },
            },
    };

    /*Stuff Related to DataTable, Actions Buttons*/
    const actionsButtonHtmlTemplate = `<div class="pad-actions-buttons">
                                        <button class="fa fa-edit  pad-actions-edit"/>
                                        <button class="fa fa-share pad-actions-preview"/>
                                        <button class="fa fa-trash pad-actions-delete"/>
                                       </div>`;

    const editPadHandler = function(event, cell){
        const {editPad, onEditPad} = props;
        const { row } = cell;
        const { id, padType } = this.getItem(row);

        onEditPad(id, padType, editPad);

        return false;
    };

    const previewPadHandler = function(event, cell){
        const { previewPad } = props;
        const { row } = cell;
        const { id } = this.getItem(row);

        previewPad(id);
        return false;
    };

    const deletePadHandler = function(event, cell){
        const { deletePad } = props;
        const { row } = cell;
        const { id, name } = this.getItem(row);

        webix.modalbox({
            type:"confirm",
            title:"Delete Pad",
            text:`Are you sure you want to permanently delete ${name}?`,
            buttons:["Cancel","Delete","x"],
        })
            .then((result)=>{

                if(result === '1'){ 
                    deletePad(id); //String '1' 
                }

            });

            return false;
    };

    const padsListClickHandlers = {
        'pad-actions-edit' : editPadHandler,
        'pad-actions-preview' : previewPadHandler,
        'pad-actions-delete' : deletePadHandler,
    };

    const getConfig = ()=>{

        const config = {
            view: 'datatable',
            css:"pads-list noselect webix_header_border webix_data_border", 
            height,
            rowHeight: 30,
            scroll: 'y',
            sorting: true,
            onClick : padsListClickHandlers,
            columns:[
                { 
                    id:"name",
                    header:{
                        text: "Name",
                        css: "sortable-column-header",
                    },
                    fillspace:true,
                    sort:"string",
                },
                { 
                    id:"modified",
                    header:{
                        text: "Modified Date",
                        css: "sortable-column-header",
                    },
                    width:170,
                    format:webix.i18n.longDateFormatStr,
                    map:"(date)#modified#", //map to coerce type for sort
                    sort:"date",
                },
                { 
                    id:"padType",
                    header:{
                        text: "Type",
                        css: "sortable-column-header",
                    },
                    width:240,
                    map:"#padType.type#", //map to get sub property
                    sort:"string",
                },
                { 
                    id:"actions",
                    header:{ 
                        text: "Actions",
                        css:"actions-column-header",
                    },
                    width:220,
                    template: actionsButtonHtmlTemplate,
                },
            ],
            
        };

        //add pager if applicable
        if(shouldPaginate === true){
            config.pager = pagerConfig;
        }
        
        return config;

    };
    

    const init = (helper) => {

            //Wipe out Pager
            const pager = document.getElementById(pagerId);
            if(pager){
                pager.innerHTML = '';
            }

            return webix.ui({
                ...getConfig(),
                container: helper.ref.current,
            });
    };

    return(
        <React.Fragment>

            <WebixHelper config={{pagination}} data={padsData} webixInitFn={init} />

           { shouldPaginate && (

                <div className="pagerBar">
                    <div className="pager" id={pagerId} ></div>
                    <div className="pagesSelect">
                            <Dropdown
                                id={`${pagerId}_Dropdown`}
                                placeholder={`${pagination}`}
                                onSelect={changePagination}
                                width={80}
                                options={pagerValues.map((e)=>`${e}`)}
                            />
                            <span>Items per page</span>

                    </div>
                    <div className="pagesListInfo"></div>
                </div>

            )}


        </React.Fragment>
        );

};


PadsList.defaultProps = {
    padsData:[],
    pagination:10,
    height:330,
    changePagination:( )=>{ },
    shouldPaginate :false,
};

PadsList.propTypes = {
    padsData: PropTypes.array,
    pagination: PropTypes.number,
    height: PropTypes.number,
    changePagination : PropTypes.func,
    shouldPaginate : PropTypes.bool,
    editPad : PropTypes.func.isRequired,
    previewPad : PropTypes.func.isRequired,
    deletePad : PropTypes.func.isRequired,
    onEditPad : PropTypes.func.isRequired,

};

export default PadsList;