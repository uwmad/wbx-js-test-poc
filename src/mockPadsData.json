[{
	"config": "0 string for config",
	"id": 0,
	"isPublic": true,
	"modified": "2000,7,25",
	"name": "0 Pad Name",
	"padType": {
		"id": 0,
		"type": "Create a Car"
	},
	"user": "0 string for user"
}, {
	"config": "1 string for config",
	"id": 1,
	"isPublic": true,
	"modified": "2001,7,25",
	"name": "1 Pad Name",
	"padType": {
		"id": 0,
		"type": "Trend Pad"
	},
	"user": "1 string for user"
}, {
	"config": "2 string for config",
	"id": 2,
	"isPublic": true,
	"modified": "2002,7,25",
	"name": "2 Pad Name",
	"padType": {
		"id": 0,
		"type": "Value Set"
	},
	"user": "2 string for user"
}, {
	"config": "3 string for config",
	"id": 3,
	"isPublic": true,
	"modified": "2003,7,25",
	"name": "3 Pad Name",
	"padType": {
		"id": 0,
		"type": "List Pad"
	},
	"user": "3 string for user"
}, {
	"config": "4 string for config",
	"id": 4,
	"isPublic": true,
	"modified": "2004,7,25",
	"name": "4 Pad Name",
	"padType": {
		"id": 0,
		"type": "Trend Pad"
	},
	"user": "4 string for user"
}, {
	"config": "5 string for config",
	"id": 5,
	"isPublic": true,
	"modified": "2005,7,25",
	"name": "5 Pad Name",
	"padType": {
		"id": 0,
		"type": "Create a Car"
	},
	"user": "5 string for user"
}, {
	"config": "6 string for config",
	"id": 6,
	"isPublic": true,
	"modified": "2006,7,25",
	"name": "6 Pad Name",
	"padType": {
		"id": 0,
		"type": "List Pad"
	},
	"user": "6 string for user"
}, {
	"config": "7 string for config",
	"id": 7,
	"isPublic": true,
	"modified": "2007,7,25",
	"name": "7 Pad Name",
	"padType": {
		"id": 0,
		"type": "Create a Car"
	},
	"user": "7 string for user"
}, {
	"config": "8 string for config",
	"id": 8,
	"isPublic": true,
	"modified": "2008,7,25",
	"name": "8 Pad Name",
	"padType": {
		"id": 0,
		"type": "List Pad"
	},
	"user": "8 string for user"
}, {
	"config": "9 string for config",
	"id": 9,
	"isPublic": true,
	"modified": "2009,7,25",
	"name": "9 Pad Name",
	"padType": {
		"id": 0,
		"type": "List Pad"
	},
	"user": "9 string for user"
},
{
	"config": "10 string for config",
	"id": 10,
	"isPublic": true,
	"modified": "2010,7,25",
	"name": "10 Pad Name",
	"padType": {
		"id": 0,
		"type": "List Pad"
	},
	"user": "10 string for user"
},
{
	"config": "11 string for config",
	"id": 11,
	"isPublic": true,
	"modified": "2011,7,25",
	"name": "11 Pad Name",
	"padType": {
		"id": 0,
		"type": "List Pad"
	},
	"user": "11 string for user"
},
{
	"config": "12 string for config",
	"id": 12,
	"isPublic": true,
	"modified": "2012,7,25",
	"name": "12 Pad Name",
	"padType": {
		"id": 0,
		"type": "List Pad"
	},
	"user": "12 string for user"
}
]