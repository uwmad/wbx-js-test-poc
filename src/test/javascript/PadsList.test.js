import React from 'react';
import { mount } from 'enzyme';
import PadsList from '../../PadsList';
import mockPadsData from '../../mockPadsData.json';

import waitForExpect from 'wait-for-expect';
import webixFromSelector from '../test_utils/webixFromSelector';


describe('PadsList', ()=>{

    const App = document.getElementById('App');

    const editPad = jest.fn();
    const onEditPad = jest.fn();
    const previewPad = jest.fn();
    const deletePad = jest.fn();
    const changePagination = jest.fn();
    const webixUISpy = jest.spyOn(webix, 'ui');


    describe('Rendering', () => {

        const wrapper = mount(<PadsList 
            editPad={editPad}
            onEditPad={onEditPad}
            previewPad={previewPad}
            deletePad={deletePad}
            padsData={mockPadsData}
            pagination={10}
            changePagination={changePagination}
            shouldPaginate={true}
            height = {700}
        />, { 
            attachTo: App
        });

        it('React OK', ()=>{
            expect(wrapper.exists()).toBe(true);
        });

        it('Webix OK', async ()=>{
            //Datatable and Dropdown
            expect(webix.ui).toHaveBeenCalledTimes(2); 
            
            const dataTable = webixFromSelector('.webix_dtable');

            expect(dataTable).toBeTruthy();
            expect(dataTable.$view.innerHTML).toMatchSnapshot();
  
            //expect({'COLUMNS':dataTable._columns}).toMatchSnapshot();
            //expect({'DATA':dataTable.data}).toMatchSnapshot();

           /* await waitForExpect(() => {

                const element = document.querySelector('.webix_column');
                expect(element).toBeTruthy();

            });*/


        });

        it('renders Webix pageDropdown', async ()=>{

            const pageDropdown = webixFromSelector('.webix_control.webix_el_combo');
            expect(pageDropdown).toBeTruthy();
            expect(pageDropdown.$view.innerHTML).toMatchSnapshot(); 

        });

        it('renders Webix pager', async ()=>{

            const pager = webixFromSelector('.webix_pager');
            expect(pager).toBeTruthy();

            //dom written by custom js
            const pagesListInfo = document.querySelector('.pagerBar .pagesListInfo');
            expect(pagesListInfo).toBeTruthy();
            expect(pagesListInfo.innerHTML).toMatchSnapshot();

        });

    });//Rendering












});//PadsList


        //webixUISpy.mockRestore();
       // wrapper.unmount();