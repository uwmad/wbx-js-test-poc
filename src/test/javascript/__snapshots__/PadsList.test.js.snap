// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PadsList Rendering Webix OK 1`] = `"<div class=\\"webix_ss_header\\" style=\\"height: 43px;\\"><div class=\\"webix_hs_left\\"></div><div class=\\"webix_hs_center\\"><table role=\\"presentation\\" style=\\"width:730px\\" cellspacing=\\"0\\" cellpadding=\\"0\\"><tbody><tr class=\\"webix_size_row\\"><td style=\\"width:100px;\\"></td><td style=\\"width:170px;\\"></td><td style=\\"width:240px;\\"></td><td style=\\"width:220px;\\"></td></tr><tr section=\\"header\\" role=\\"row\\"><td role=\\"presentation\\" column=\\"0\\" class=\\"webix_first webix_last_row\\"><div role=\\"columnheader\\" class=\\"webix_hcell sortable-column-header\\">Name</div></td><td role=\\"presentation\\" column=\\"1\\" class=\\" webix_last_row\\"><div role=\\"columnheader\\" class=\\"webix_hcell sortable-column-header\\">Modified Date</div></td><td role=\\"presentation\\" column=\\"2\\" class=\\" webix_last_row\\"><div role=\\"columnheader\\" class=\\"webix_hcell sortable-column-header\\">Type</div></td><td role=\\"presentation\\" column=\\"3\\" class=\\" webix_last webix_last_row\\"><div role=\\"columnheader\\" class=\\"webix_hcell actions-column-header\\">Actions</div></td></tr></tbody></table></div><div class=\\"webix_hs_right\\"></div></div><div class=\\"webix_ss_body\\"><div class=\\"webix_ss_left\\" style=\\"height: 657px;\\"><div class=\\"webix_ss_center_scroll\\" style=\\"height: 657px;\\"></div></div><div class=\\"webix_ss_center\\" style=\\"height: 657px;\\"><div class=\\"webix_ss_center_scroll\\" role=\\"rowgroup\\" style=\\"height: 657px;\\"></div></div><div class=\\"webix_ss_right\\" style=\\"height: 657px;\\"><div class=\\"webix_ss_center_scroll\\" style=\\"height: 657px;\\"></div></div></div><div class=\\"webix_ss_hscroll webix_vscroll_x\\" role=\\"scrollbar\\" aria-orientation=\\"horizontal\\"><div class=\\"webix_vscroll_body\\" style=\\"width: 730px;\\"></div></div><div class=\\"webix_ss_footer\\"><div class=\\"webix_hs_left\\"></div><div class=\\"webix_hs_center\\"></div><div class=\\"webix_hs_right\\"></div></div><div class=\\"webix_ss_vscroll_header\\" style=\\"height: 42px; width: 16px;\\"></div><div class=\\"webix_ss_vscroll webix_vscroll_y\\" role=\\"scrollbar\\" aria-orientation=\\"vertical\\" style=\\"display: block; margin-top: 43px; height: 657px; width: 17px;\\"><div class=\\"webix_vscroll_body\\" style=\\"height: 300px;\\"></div></div><div class=\\"webix_ss_vscroll_footer\\"></div>"`;

exports[`PadsList Rendering renders Webix pageDropdown 1`] = `"<div class=\\"webix_el_box\\" style=\\"width:80px; height:40px\\"><input role=\\"combobox\\" placeholder=\\"10\\" id=\\"1546322400090\\" type=\\"combo\\" value=\\"\\" style=\\"width: 76px; text-align: left;\\" aria-autocomplete=\\"list\\" aria-expanded=\\"false\\"><span style=\\"height:27px;padding-top:7px;\\" class=\\"webix_input_icon wxi-angle-down\\"></span></div>"`;

exports[`PadsList Rendering renders Webix pager 1`] = `"1 to 10 of 13 "`;
